### ISSUE TITLE
Running Lilotane on classical planning problems

### ISSUE THREAD

**Issue by bo198214** [domschrei/lilotane#8]
lilotane is dumping core with the following domain and problem:
```
(define (domain trivial01-gen)
  (:requirements :strips :typing)
  (:types 
         xloc yloc tile - object
  )
  (:predicates (adjwe ?h1 - xloc ?h2 - xloc) (adjns ?v1 - yloc ?v2 - yloc) 
        (at ?t - tile ?h - xloc ?v - yloc) (empty ?h - xloc ?v - yloc) 
  )
  (:action move
   :parameters (?t - tile ?x0 - xloc ?x1 - xloc ?y1 - yloc)
   :precondition (and
        (at ?t ?x1 ?y1)
        (adjwe ?x0 ?x1)
        (empty ?x0 ?y1)
    )
   :effect (and 
        (not (at ?t ?x1 ?y1)) (empty ?x1 ?y1)
        (at ?t ?x0 ?y1) (not (empty ?x0 ?y1))
    )
  )
)
```
```
(define (problem trivial01-gen)
    (:domain trivial01-gen)
    (:objects 
        h1 - xloc h2 - xloc
        v1 - yloc
        tile1 - tile
    )
    (:init 
        (adjwe h1 h2)
        (empty h1 v1) (at tile1 h2 v1)
    )
    (:goal (and (at tile1 h1 v1) (empty h2 v1)))
)
```

Other planner do solve the problem, so here I am out of ideas, what it could be.

**Comment by domschrei** [comment]
Hi there,

your example problem is a classical planning problem. Lilotane is a **hierarchical** planner, more specifically: a Total-order Hierarchical Task Network (TOHTN) planner. You can find a precise definition of Lilotane's problem model in Section 2.1 of [this paper](https://jair.org/index.php/jair/article/view/12520/26669). The input language for (TO)HTN planning problems, HDDL, is described [here](https://ojs.aaai.org/index.php/AAAI/article/download/6542/6398).

In short, in order to run Lilotane on a planning problem, you need to specify an initial task network (`:htn` struct in problem file) and any methods necessary to decompose it (`:method` struct in domain file). You can find lots of examples in the `instances/` directory of this repository.

**Comment by bo198214** [comment]
Thank you for your fast reply. I was aware that there is a difference between PDDL and HDDL, however (without diving into the details) I thought HDDL is an extension of PDDL. I am just a hobbyist, I have no deep knowledge about planners. So before I really reading all the papers and understanding it (which may take a long time due to lacking background knowledge), can you just tell me (thought it's ot of scope of this "issue") can I transform each classical problem into a TOHTN? (if yes I can find my way through how to do this on my own)

**Comment by domschrei** [comment]
I see.

>can I transform each classical problem into a TOHTN?

Short answer: Yes, but it is most likely inefficient (compared to "actual" hierarchical domains) if done naïvely.

See, hierarchical planning depends on expert knowledge on the planning domain at hand which you do not have (or do not make use of) in classical planning. This expert knowledge is your task network and its possible decompositions. If you want to make your domain hierarchical, you can always use the following transformation:

* Introduce a single abstract task `decompose()`. All methods you introduce belong to this task.
* For every operator `op(·)` in your domain, introduce a method `m_op(·)` (with all arguments of `op`) with the two subtasks `op(·)` (execute the action) and `decompose()`.
* Introduce one more method `m_stop()` without any subtasks.
* As your initial task network, pick the single task `decompose()`.

Now every time the planner picks a decomposition, it can either end the task network by picking `m_stop` or execute a single applicable action `op` with `m_op` and continue the loop. If the classical planning problem can be solved by some sequence of actions, then an HTN planning procedure can find a series of decompositions as well in order to transform the initial state into a goal state.

This is likely an inefficient way to create a hierarchical domain because you do not actually incorporate any knowledge of your domain, and the "structure" you induce is just a linear chain of arbitrary actions. In Lilotane, the effort spent on a problem increases considerably with the depth of the network at hand, and in such a case the depth is proportional to the plan length. However, if you feel like it, try it out, maybe it's not that bad! If you add a second instance of `decompose()` to the subtasks of each `m_op`, you could achieve a transformation where the depth of your network will be logarithmic in the found plan length, which might be better in the case of Lilotane.

A better way to create a hierarchical domain from a classical domain would be to take a step back and think less mechanically. Rather: How could you model the problem in a hierarchical manner, i.e., how can you recursively break down the logic of your planning problem? There is no easy answer here, even less so without knowing the particular problem you try to solve. 

There has also been some research going into automatically inferring useful HTN models from a classical planning domain, but I don't have a link right now.

**Comment by bo198214** [comment]
Unfortunately in the domain I want to use the solver for, not much is known - it's sliding puzzles. There are some combinatorial ways to find whether a solution exists in the case of 8puzzle or 15puzzle. But apart from that I am not aware that there are useful sub-problems or something like that.
So would you rather discourage using lilotane for this kind of problems?
Anyways I will try it out (just to learn HDDL) and in some time report my findings here.

**Comment by domschrei** [comment]
Yes, sliding puzzles are generally not so well-suited for hierarchical planning. If there is an algorithm to solve a sliding puzzle, you can sort of implement it via methods and subtasks, but this is not the actual purpose of HTN planning – the original idea of HTNs is to express the logic of your problem, not the algorithm to solve it, in a hierarchical manner. And the mechanics of sliding puzzles are remarkably non-hierarchical, I would say :) Classical planners are the more natural choice there.

If you want to try out the above compilation, I'd be happy to see how well Lilotane fares against classical planners on "their" domain(s)! Would be interesting.

**Comment by bo198214** [comment]
Somehow it seems I am doing still something wrong, lilotane will not find a solution do this hddl until layer 898 (and then gives up, probably due to memory restrictions).
But this puzzle has definitively a solution which is found rather quickly with a classical planner (fast downward on the corresponding pddl).
So what am I doing wrong (dont let the strange names bother you, its made with a generator)?
```
(define (domain n8puzzle01-hddl-gen)
  (:requirements :strips :typing)
  (:types 
         xloc yloc tile - object
		 type_1 - tile
  )
  (:predicates (adjwe ?h1 - xloc ?h2 - xloc) (adjns ?v1 - yloc ?v2 - yloc) 
        (at ?t - tile ?h - xloc ?v - yloc) (empty ?h - xloc ?v - yloc) 
  )
  (:task decompose
   :parameters ()
  )
  (:method m_stop
   :task (decompose)
  )    

  (:method m_move-type_1-s
   :parameters (?t - type_1 ?x1 - xloc ?y0 - yloc ?y1 - yloc)
   :task (decompose)
   :subtasks (and
       (move-type_1-s ?t ?x1 ?y0 ?y1)
       (decompose)
    )
   )
  (:action move-type_1-s
   :parameters (?t - type_1 ?x1 - xloc ?y0 - yloc ?y1 - yloc)
   :precondition (and
        (at ?t ?x1 ?y1)
        
        (adjns ?y0 ?y1)
        (empty ?x1 ?y0)
    )
   :effect (and 
        (not (at ?t ?x1 ?y1)) (empty ?x1 ?y1)
        (at ?t ?x1 ?y0) (not (empty ?x1 ?y0))
    )
  )
  (:method m_move-type_1-n
   :parameters (?t - type_1 ?x1 - xloc ?y1 - yloc ?y2 - yloc)
   :task (decompose)
   :subtasks (and
       (move-type_1-n ?t ?x1 ?y1 ?y2)
       (decompose)
    )
   )
  (:action move-type_1-n
   :parameters (?t - type_1 ?x1 - xloc ?y1 - yloc ?y2 - yloc)
   :precondition (and
        (at ?t ?x1 ?y1)
        
        (adjns ?y1 ?y2)
        (empty ?x1 ?y2)
    )
   :effect (and 
        (not (at ?t ?x1 ?y1)) (empty ?x1 ?y1)
        (at ?t ?x1 ?y2) (not (empty ?x1 ?y2))
    )
  )
  (:method m_move-type_1-e
   :parameters (?t - type_1 ?x1 - xloc ?x2 - xloc ?y1 - yloc)
   :task (decompose)
   :subtasks (and
       (move-type_1-e ?t ?x1 ?x2 ?y1)
       (decompose)
    )
   )
  (:action move-type_1-e
   :parameters (?t - type_1 ?x1 - xloc ?x2 - xloc ?y1 - yloc)
   :precondition (and
        (at ?t ?x1 ?y1)
        (adjwe ?x1 ?x2)
        
        (empty ?x2 ?y1)
    )
   :effect (and 
        (not (at ?t ?x1 ?y1)) (empty ?x1 ?y1)
        (at ?t ?x2 ?y1) (not (empty ?x2 ?y1))
    )
  )
  (:method m_move-type_1-w
   :parameters (?t - type_1 ?x0 - xloc ?x1 - xloc ?y1 - yloc)
   :task (decompose)
   :subtasks (and
       (move-type_1-w ?t ?x0 ?x1 ?y1)
       (decompose)
    )
   )
  (:action move-type_1-w
   :parameters (?t - type_1 ?x0 - xloc ?x1 - xloc ?y1 - yloc)
   :precondition (and
        (at ?t ?x1 ?y1)
        (adjwe ?x0 ?x1)
        
        (empty ?x0 ?y1)
    )
   :effect (and 
        (not (at ?t ?x1 ?y1)) (empty ?x1 ?y1)
        (at ?t ?x0 ?y1) (not (empty ?x0 ?y1))
    )
  )
)
```
```
(define (problem n8puzzle01-hddl-gen)
    (:domain n8puzzle01-hddl-gen)
    (:objects 
        h1 - xloc h2 - xloc h3 - xloc
        v1 - yloc v2 - yloc v3 - yloc
        tile1 - type_1 tile2 - type_1 tile3 - type_1 tile4 - type_1 tile5 - type_1 tile6 - type_1 tile7 - type_1 tile8 - type_1
    )
    (:htn
     :tasks (decompose)
    )
    (:init 
        (adjwe h1 h2) (adjwe h2 h3)
        (adjns v1 v2) (adjns v2 v3)

        (empty h1 v3) (at tile8 h2 v3) (at tile7 h3 v3)
        (at tile6 h1 v2) (at tile5 h2 v2) (at tile4 h3 v2)
        (at tile3 h1 v1) (at tile2 h2 v1) (at tile1 h3 v1)

    )
    (:goal (and (at tile1 h1 v3) (at tile2 h2 v3) (at tile3 h3 v3) (at tile4 h1 v2) (at tile5 h2 v2) (at tile6 h3 v2) (at tile7 h1 v1) (at tile8 h2 v1) (empty h3 v1)))
)
```

**Comment by domschrei** [comment]
Indeed there seems to be a problem on the master branch right now. I need to investigate this further. 

In the meantime, please switch to the more recent `factframe` branch where your problem appears to work fine after a small bugfix from my side.

**Comment by bo198214** [comment]
This works!
Was the hierarchical setup too trivial, that it didn't work out of the box?

Just FYI: the fast downward solved the corresponding pddl in 14s with the same result as lilotane, which solved it in 38s.
I would say, that's not bad, for being focussed on the HTN!

**Comment by domschrei** [comment]
I am still unsure what was the problem. However, in the `master` branch your problem does work if precondition inference is switched off (`-mp=0`), so there is likely a problem in the instantiation part of the master branch. We will probably merge an updated version into master soon.

That actually sounds not too bad. For a true performance comparison one would probably need to test it on more (easier and harder) problems and see to which degree the run time profiles diverge. Thanks for letting me know!

If you feel like improving your hierarchical domain a little further, you could try to reduce the required depth to find a plan using the logarithmic technique I mentioned above.
However this would probably only work well in conjunction with a concious effort to keep search space as small as possible. You could introduce an additional predicate `(done)` which is switched on by the method `m_stop` (i.e., by a child action thereof), and then enforce that "proper" actions can only be executed if `(done)` does not hold. This way you tell the solver exactly where the proper actions need to go (on the chronologically leftmost part of a layer, that is), and the internal SAT solver does no unnecessary branching over where to put the needed `stop` methods.
